### PR TITLE
Get rid of frontend compat patchinfo

### PR DESCRIPTION
**Pull request by bgeuken**

### CODE REVIEW

File: src/api/app/controllers/webui/patchinfo_controller.rb [hunk 1]
```diff
@@ -193,14 +193,43 @@ def save
   end
 
   def remove
+    # checks
+    if @package.name == '_project'
+      raise DeletePackageNoPermission.new '_project package can not be deleted.'
+    end
+
+    unless User.current.can_modify_package?(@package)
```
Comment (ChrisBr) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: I'd prefer to use
```
authorize package :new_incident?
```

(probably you need to introduce a policy for package ...)

File: src/api/app/controllers/webui/patchinfo_controller.rb [hunk 2]
```diff
@@ -193,14 +193,43 @@ def save
...
+      raise DeletePackageNoPermission.new "no permission to delete package #{@package.name} in project #{@project.name}"
+    end
+
+    # deny deleting if other packages use this as develpackage
+    # Note: This might raise an exception
+    @package.can_be_deleted?
```
Comment (ChrisBr) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: I'd also check this in a policy

File: src/api/app/controllers/webui/patchinfo_controller.rb [hunk 3]
```diff
@@ -193,14 +193,43 @@ def save
...
+      Package.transaction do
+        # we need to keep this order to delete first the api model
+        @package.revoke_requests
+        @package.destroy
+
+        path << Suse::Backend.build_query_from_hash(parameters, [:user, :comment])
+        Suse::Backend.delete(path)
+      end
+    rescue ActiveXML::Transport::Error, ActiveXML::Transport::NotFoundError => e
       flash[:error] = e.summary
```
Comment (ChrisBr) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: Don't you need to redirect here, otherwise your flash will be overwritten by the flash[:notice]

File: src/api/app/controllers/webui/patchinfo_controller.rb [hunk 4]
```diff
@@ -193,14 +193,43 @@ def save
...
-    rescue ActiveXML::Transport::Error => e
+      Package.transaction do
+        # we need to keep this order to delete first the api model
+        @package.revoke_requests
+        @package.destroy
+
+        path << Suse::Backend.build_query_from_hash(parameters, [:user, :comment])
+        Suse::Backend.delete(path)
+      end
+    rescue ActiveXML::Transport::Error, ActiveXML::Transport::NotFoundError => e
```
Comment (ChrisBr) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: I thought we want to get rid of ActiveXML at all? Does the backend throw ActiveXML exceptions?

Comment (bgeuken) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: One thing at a time;-)
And yes [it does](https://github.com/openSUSE/open-build-service/blob/master/src/api/lib/opensuse/backend.rb#L173).

File: src/api/app/controllers/webui/patchinfo_controller.rb (hunk 3 above)
Comment (coolo) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: there are 2 pieces to activexml that should be seperated I guess. One is the XML parsing and one is the complete backend handling (ActiveXML::Transport).

File: src/api/app/controllers/webui/patchinfo_controller.rb (hunk 4 above)
Comment (hennevogel) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: Before we merged API/WebUI it was:
*Webui <- ActiveXML -> API <- ActiveXML -> Backend* 

Now it is:
*API (Webui Namespace) <- ActiveXML -> API <- ActiveXML -> Backend*

We want it to be:
*API <- ActiveXML -> Backend*

File: src/api/app/controllers/webui/patchinfo_controller.rb [hunk 5]
```diff
@@ -193,14 +193,43 @@ def save
...
+      package: @package.name,
+      timeout: 500
+    }
     begin
-      FrontendCompat.new.delete_package project: @project, package: @package
-      flash[:notice] = "'#{@package}' was removed successfully from project '#{@project}'"
-      Rails.cache.delete('%s_packages_mainpage' % @project)
-      Rails.cache.delete('%s_problem_packages' % @project)
-    rescue ActiveXML::Transport::Error => e
+      Package.transaction do
```
Comment (hennevogel) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: This should go to Package. Similar to Project.delete_on_backend

File: src/api/app/controllers/webui/patchinfo_controller.rb [hunk 6]
```diff
@@ -193,14 +193,43 @@ def save
...
+        # we need to keep this order to delete first the api model
+        @package.revoke_requests
```
Comment (hennevogel) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: This should go to Package. Similar to Project.cleanup_before_destroy

File: src/api/app/controllers/webui/patchinfo_controller.rb [hunk 7]
```diff
@@ -193,14 +193,43 @@ def save
   end
 
   def remove
+    # checks
+    if @package.name == '_project'
+      raise DeletePackageNoPermission.new '_project package can not be deleted.'
```
Comment (hennevogel) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: Why do you raise an exception here?

File: src/api/app/controllers/webui/patchinfo_controller.rb (hunk 1 above)
Comment (bgeuken) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: That's a good idea. Though I would prefer to do that in a separate pull request.

File: src/api/app/controllers/webui/patchinfo_controller.rb (hunk 2 above)
Comment (hennevogel) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: I wouldn't. This is has nothing to do with the permission of the user to delete this package. This is some functional requirement for the application

File: src/api/app/controllers/webui/patchinfo_controller.rb (hunk 1 above)
Comment (ChrisBr) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: Agreed!

File: src/api/app/controllers/webui/patchinfo_controller.rb (hunk 2 above)
Comment (ChrisBr) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: Hm, okay!

File: src/api/app/controllers/webui/patchinfo_controller.rb (hunk 5 above)
Comment (bgeuken) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: @hennevogel For now this is just a simple replacement for using the [api source controller](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/controllers/source_controller.rb#L291). So the code is pretty similar to what is done there.
My plan was to adjust the code in a couple of follow up pull requests, including a cleanup of the current tests.

Maybe it would be better to close the PR and submit the changes all together

Comment (hennevogel) [inline src/api/app/controllers/webui/patchinfo_controller.rb]: Sounds reasonable

File: src/api/test/functional/webui/patchinfo_create_test.rb [hunk 8]
```diff
@@ -120,7 +120,17 @@ def delete_patchinfo project
...
+    assert_equal page.current_path, project_show_path(project)
     find('#flash-messages').must_have_text "'patchinfo' was removed successfully from project"
+
+    # FIXME: There must be a better way to test this
+    begin
+      Package.get_by_project_and_name(project.to_param, "patchinfo")
+      assert false
+    rescue Package::UnknownObjectError => e
+      assert_equal "home:Iggy/patchinfo", e.message
+    end
```
Comment (adrianschroeter) [inline src/api/test/functional/webui/patchinfo_create_test.rb]: You can use:

exception = assert_raise Package::UnknownObjectError do
...
end
assert_equal "home:Iggy/patchinfo", exception.message
